README.md
# Dedupr

Quick and dirty Node.js tool to find (and delete) duplicate files. Not the fastest, not the most featured, not the most popular out there, and no voodoo involved. But it works well enough, and can be used directly via the command line or imported as a library on your Node.js app.

## Features

-   Can be used via command line or imported as a library
-   Find duplicates based on file size, content hash and filenames (optional)
-   Supports most common hashing algorithms (SHA1, SHA256, SHA512, MD4, MD5 etc)
-   Delete duplicates automatically (optional)
-   Results are exported to a JSON file

## How to install

To install globally on your machine please use:

    $ npm install dedupr -g

Or to install locally on your current project:

    $ npm install dedupr --save

## Command line usage

    $ dedupr -[options] folders

Detect duplicates on some user folders, using defaults:

    $ dedupr /home/joe /home/karen /home/sara

Detect duplicates on all user's home folders, verbose mode activated:

    $ dedupr -v /home

Delete duplicates on logged user's home folder, using the "fast" hashing preset:

    $ dedupr -d --fast ~/

Delete duplicate images considering filenames, reverse order, MD5 hashing only the first and last 1KB of files (hash size 2KB):

    $ dedupr -o /home/custom-dedupr-output.json \
             -e jpg gif png bmp \
             -s 2
             -h md5
             -f -r -d
             ~/photos ~/camera ~/pictures ~/downloads

## Importing as a library

```typescript
import Dedupr from "dedupr"
// const Dedupr = require("dedupr").default

const options = {
    console: true,
    folders: ["/home/user1/photos", "/home/user2/photos", "/home/user3/photos"],
    hashAlgorithm: "sha1"
}

const dedupr = new Dedupr(options)
await dedupr.run()

console.dir(dedupr.results)
```

## Options

### console

Enable or disable logging to the console. Enabled by default when using via the command line, but not when using it as a library / programatically.

### folders

List of folders that should be scanned. On the command line, these are the last arguments to be passed. If any of these folders do not exist, the tool will throw an exception.

By default, duplicates will be detected based on alphabetical / ascending order. If you pass /folderA, folderB and folderC, in that order, duplicates will be flagged on folderB and folderC only. If a file is present on folderB and folderC, the one inside folderC will be flagged. This behaviour can be changed with the `reverse` option.

### extensions _`-e`_

Array of file extensions that should be included. Defaults to all files.

### output _`-o`_

Save results to the specified output file. Defaults to `dedupr.json` on the current folder. If you want to disable saving the output, set this to `false`

### reverse _`-r`_

Sort folders and files from Z to A to 0 (descending). By default, they are sorted alphabetically (ascending). Please note that this also changes the order of the passed folders, so the very last occurrence of a file will be the non-duplicate.

### filename _`-f`_

In addition to the hash value and file size, also consider the filename to find duplicates. Meaning files will identical contents but different filenames won't be marked as duplicates. Default is `false`.

### verbose _`-v`_

Activate verbose mode with extra logging. Defaults to `false`.

### delete _`-d`_

Delete duplicates. Only the very first occurrence of a file will remain (or very last, if `reverse` is set). **Use with caution!**

## Advanced options

### parallel _`-p`_

How many files should be hashed in parallel. Defaults to `3`.

### hashSize _`-s`_

How many bytes should be hashed from each file? Defaults to `2048`, meaning it uses the first and last 1MB of a file. Depending on your use case and the available CPU power, you might want to reduce this value.

### hashAlgorithm _`-h`_

Which hashing algorithm should be used. Default is `sha256`. Some of the other possible values: "sha1", "sha512", "md5", "blake2b512".

## Shortcut options

**Please note that the options below will always override the `hashSize` and `hashAlgorithm` values.**

### _`--crazyfast`_

Same as `-s 4 -h sha1`. Hashes the first and last 2KB of files, using SHA1. Use with caution, as this _might_ catch some false positives.

### _`--veryfast`_

Same as `-s 64 -h sha1`. Hashes the first and last 32KB of files, using SHA1.

### _`--faster`_

Same as `-s 512 -h sha1`. Hashes the first and last 256KB of files, using SHA1.

### _`--fast`_

Same as `-s 1024 -h sha256`. Hashes the first and last 512KB of files, using SHA256.

### _`--safe`_

Same as `-s 32768 -h sha256`. Hashes the first and last 16MB of files, using SHA256. Might be a bit slow if you have many media files.

### _`--safer`_

Same as `-s 131072 -h sha512`. Hashes the first and last 64MB of files, using SHA512. Very slow if you have many media files.

## Need help?

Post an issue [here](https://github.com/igoramadas/dedupr/issues).


History.md
# Changelog for Dedupr

2.0.2
=====
* Updated dependencies, fixing the occasional "cannot find module util/types" bug.

2.0.1
=====
* Better handling of unsynchronized (cloud / virtual) files. (by @noomorph)
* File permissions issue fixed. (by @noomorph)
* Updated dependencies.

2.0.0
=====
* Full rewrite of the tool using TypeScript, now renamed to Dedupr.
* NEW! Options to set the hash size and algorithm.

1.0.0
=====
* Initial release (previously named Dedup.js).


src/cmd.ts
// COMMAND LINE WRAPPER

import {hasValue, logError} from "./utils"
import Dedupr from "./index"
import yargs from "yargs"
import yargsIntance from "yargs/yargs"

// Unhandled rejections goes here.
process.on("unhandledRejection", (err) => {
    console.error("FATAL ERROR!")
    console.error(err)
    return process.exit()
})

export = async function () {
    const argOptions = yargsIntance(process.argv.slice(2)).options({
        e: {alias: "extensions", type: "array", describe: "Allowed file extensions, default is all extensions"},
        o: {alias: "output", type: "string", describe: "Full path to the JSON output file, default is dedupr.json"},
        v: {alias: "verbose", type: "boolean", describe: "Verbose mode with extra logging"},
        r: {alias: "reverse", type: "boolean", describe: "Reverse the folders and files order (alphabetically descending)"},
        f: {alias: "filename", type: "boolean", describe: "Also consider filenames to check if a file is a duplicate"},
        d: {alias: "delete", type: "boolean", describe: "Delete duplicate files"},
        p: {alias: "parallel", type: "number", describe: "How many files processed in parallel (default 3)"},
        s: {alias: "size", type: "number", describe: "How much data (kilobytes) to hash from start and end of each file"},
        h: {alias: "hash", type: "string", describe: "Hash algorithm, default is sha1"},
        crazyfast: {type: "boolean", describe: "Shortcut to --size 4, --hash sha1"},
        veryfast: {type: "boolean", describe: "Shortcut to --size 64, --hash sha1"},
        faster: {type: "boolean", describe: "Shortcut to --size 512, --hash sha1"},
        fast: {type: "boolean", describe: "Shortcut to --size 1024, --hash sha256"},
        safe: {type: "boolean", describe: "Shortcut to --size 32768, --hash sha256"},
        safer: {type: "boolean", describe: "Shortcut to --size 131072, --hash sha512"}
    })

    // Arguments shortcut.
    const argv = argOptions.parseSync()

    // Option grouping.
    argOptions.group(["e", "o", "r", "f", "d"], "Options:")
    argOptions.group(["p", "s", "h"], "Advanced options:")
    argOptions.group(["crazyfast", "veryfast", "faster", "fast", "safe"], "Shortcuts:")

    // Command line options.
    argOptions.env("DEDUPR")
    argOptions.wrap(Math.min(100, yargs.terminalWidth()))

    // Examples.
    argOptions.usage(`Usage: $0 -[options...] folders...`)
    argOptions.example(`$ $0 --fast ~/`, "")
    argOptions.example(`$ $0 -e jpg gif png --veryfast ~/photos ~/camera ~/downloads`, "")
    argOptions.example(`$ $0 -f -d -h sha512 -o duplicate-report.json /backup`, "")
    argOptions.example(`$ $0 -h md5 -s 16 /var`, "")
    argOptions.epilog("Need help? More info at https://github.com/igoramadas/dedupr")
    argOptions.help().demandCommand(1, "")

    // Transform arguments to options.
    let options: Options = {
        console: true,
        folders: argv._ as string[],
        extensions: hasValue(argv.e) ? (argv.e as string[]) : null,
        output: hasValue(argv.o) ? argv.o : null,
        verbose: hasValue(argv.v) ? argv.v : null,
        reverse: hasValue(argv.r) ? argv.r : null,
        filename: hasValue(argv.f) ? argv.f : null,
        delete: hasValue(argv.d) ? argv.d : null,
        parallel: hasValue(argv.p) ? argv.p : null,
        hashSize: hasValue(argv.s) ? argv.s : null,
        hashAlgorithm: hasValue(argv.h) ? argv.h : null
    }

    // Hash size shortcuts (crazyfast 4KB, superfast 64KB, faster 512KB, fast 1MB, safe 32MB, safer 128MB).
    if (argv.crazyfast) {
        options.hashSize = 4
        options.hashAlgorithm = "sha1"
    } else if (argv.veryfast) {
        options.hashSize = 64
        options.hashAlgorithm = "sha1"
    } else if (argv.faster) {
        options.hashSize = 512
        options.hashAlgorithm = "sha1"
    } else if (argv.fast) {
        options.hashSize = 1024
        options.hashAlgorithm = "sha256"
    } else if (argv.safe) {
        options.hashSize = 32768
        options.hashAlgorithm = "sha256"
    } else if (argv.safer) {
        options.hashSize = 131072
        options.hashAlgorithm = "sha512"
    }

    // Do it baby!
    try {
        const dedupr = new Dedupr(options)
        await dedupr.run()
    } catch (ex) {
        logError(options, `Failure to run`, ex)
    }
}
